client/go/dogma/watch_service.go

```diff
@@ -0,0 +1,358 @@
+//
+//   https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+// WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+// License for the specific language governing permissions and limitations
+// under the License.
+
+package dogma
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"math/rand"
+	"net/http"
+	"net/url"
+	"strconv"
+	"strings"
+	"sync"
+	"sync/atomic"
+	"time"
+	"unsafe"
+)
+
+type watchService service
+
+// WatchResult represents a result from watch operation.
+type WatchResult struct {
+	Commit *Commit
+	Res    *http.Response
+	Err    error
+}
+
+func (ws *watchService) watchFile(ctx context.Context, projectName, repoName, lastKnownRevision string,
+	query *Query, timeout time.Duration) <-chan *WatchResult {
+	watchResult := make(chan *WatchResult)
+	if query == nil {
+		watchResult <- &WatchResult{Err: errors.New("query should not be nil")}
+		return watchResult
+	}
+
+	u := fmt.Sprintf("%vprojects/%v/repos/%v/contents%v", defaultPathPrefix, projectName, repoName, query.Path)
+	v := &url.Values{}
+	if query != nil && query.Type == JSONPath {
+		if err := setJSONPaths(v, query.Path, query.Expressions); err != nil {
+			watchResult <- &WatchResult{Err: err}
+			return watchResult
+		}
+	}
+	u += encodeValues(v)
+	ws.watchRequest(ctx, watchResult, u, lastKnownRevision, timeout)
+	return watchResult
+}
+
+func (ws *watchService) watchRepo(ctx context.Context,
+	projectName, repoName, lastKnownRevision, pathPattern string, timeout time.Duration) <-chan *WatchResult {
+	watchResult := make(chan *WatchResult)
+	if len(pathPattern) != 0 && !strings.HasPrefix(pathPattern, "/") {
+		// Normalize the pathPattern when it does not start with "/" so that the pathPattern fits into the url.
+		pathPattern = "/**/" + pathPattern
+	}
+
+	u := fmt.Sprintf("%vprojects/%v/repos/%v/contents%v", defaultPathPrefix, projectName, repoName, pathPattern)
+	ws.watchRequest(ctx, watchResult, u, lastKnownRevision, timeout)
+	return watchResult
+}
+
+func (ws *watchService) watchRequest(ctx context.Context, watchResult chan<- *WatchResult,
+	u, lastKnownRevision string, timeout time.Duration) {
+	req, err := ws.client.newRequest("GET", u, nil)
+	if err != nil {
+		watchResult <- &WatchResult{Err: err}
+		return
+	}
+	if len(lastKnownRevision) != 0 {
+		req.Header.Set("if-none-match", lastKnownRevision)
+	} else {
+		req.Header.Set("if-none-match", "-1")
+	}
+	if timeout != 0 {
+		req.Header.Set("prefer", fmt.Sprintf("wait=%v", timeout.Seconds()))
+	}
+
+	go func() {
+		commit := new(Commit)
+		res, err := ws.client.do(ctx, req, commit)
+		if err != nil {
+			watchResult <- &WatchResult{Commit: nil, Res: res, Err: err}
+		} else {
+			watchResult <- &WatchResult{Commit: commit, Res: res, Err: nil}
+		}
+	}()
+}
+
+const watchTimeout = 1 * time.Minute
```

Comment:
user couldn't change timeout value?

Reply:
I didn't make it because the Java version does not have the functionality. Do you think we need it? Then I think I have to change the Java version as well. :)